scalableminds/webknossos
# Support cyclic trees

This PR adds support for cyclic trees (they're not really trees any longer ^^). The warning in the abstract tree viewer regarding cyclic trees was broken, opening the tab resulted in an infinite loop that nearly crashed my system. I've added a small cycle check and render an error message if a cyclic tree was detected.
The only other place I could find that assumed non-cyclic trees was the deleteNode/deleteEdge functionality. I've fixed this as well.

We should also think about whether we want to allow to create cyclic trees from within webknossos. Currently the merge functionality will not allow this, and maybe this is fine - we could wait for the lab to request that feature.

Does anyone else know of other places in the code that assume there are no cyclic trees? (@normanrz @hotzenklotz @philippotto)

@jfrohnhofen As far as I know, the backend makes no assumptions about the absence of cycles in trees, right?

### Mailable description of changes:
 - Cyclic trees/graphs are now officially supported by webKnossos. Splitting/Merging actions on those trees will no longer break tracings.


### Steps to test:
- Upload this nml [cyclic.nml.zip](https://github.com/scalableminds/webknossos/files/1461491/cyclic.nml.zip) or create other cyclic nmls
- Split trees, undo, redo, try to break webknossos - webknossos should not break :D

### Issues:
- fixes #2130

------
